File: 6-cqrs-and-read-models/1-java/write-service/src/main/java/com/starci/write/customer/CustomerController.java
package com.starci.write.customer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Write-side REST controller.
 *
 * <p>Mirrors TS CustomerController:
 * <pre>
 *   POST /customer/update  { id, name, email }  →  upsert PostgreSQL + emit RabbitMQ event
 * </pre>
 */
@RestController
@RequestMapping("/customer")
public class CustomerController {

    private static final Logger log = LoggerFactory.getLogger(CustomerController.class);

    private final CustomerService service;

    public CustomerController(CustomerService service) {
        this.service = service;
    }

    /**
     * Upsert a customer profile (create or update) and publish an event.
     *
     * @param body request body with id, name, email
     * @return the persisted customer
     */
    @PostMapping("/update")
    public ResponseEntity<Customer> update(@RequestBody UpsertRequest body) {
        log.info("Received update request for customer \"{}\"", body.getId());
        Customer saved = service.upsert(body.getId(), body.getName(), body.getEmail());
        return ResponseEntity.ok(saved);
    }
}

File: 6-cqrs-and-read-models/write-service/src/customer/customer.controller.ts
/**
 * HTTP controller — Command side. POST /customer/update dispatches an
 * UpsertCustomerCommand on the CQRS CommandBus.
 * (VI: Controller phía ghi — POST /customer/update đẩy UpsertCustomerCommand lên CommandBus.)
 */
import { Body, Controller, Logger, Post } from '@nestjs/common';
import { CommandBus } from '@nestjs/cqrs';
import { UpsertCustomerCommand } from './commands';

@Controller('customer')
/**
 * Class `CustomerController` — thành phần lab (controller/service/module).
 * (EN: Class `CustomerController` — lesson lab component.)
 */
export class CustomerController {
    private readonly logger = new Logger(CustomerController.name);

    constructor(private readonly commandBus: CommandBus) {}

    @Post('update')
    async update(
        @Body() body: { id: string; name: string; email: string },
    ): Promise<unknown> {
        this.logger.log(`Received update request for customer "${body.id}"`);
        return this.commandBus.execute(
            new UpsertCustomerCommand(body.id, body.name, body.email),
        );
    }
}
